### PR TITLE
Trace the Gemma4 merge dtype structurally instead of by substring

### PR DESCRIPTION
**Pull request by danielhanchen**
`Core (HF=latest + TRL=latest)` is red on unsloth main. One of its four failures is this zoo guard:

```
FAILED tests/test_gemma4_dtype_drift_guards.py::test_real_gemma4_image_and_video_merges_still_cast_dtype
  assert 'image_features.to(inputs_embeds.device, inputs_embeds.dtype)' in '<modeling_gemma4.py>'
```

## This is not a functional break

`get_image_features` became variable-resolution in transformers 5.15.0, so `pooler_output` is a tuple and `forward` has to concatenate it:

```python
# 5.14.1
image_features = image_features.to(inputs_embeds.device, inputs_embeds.dtype)

# 5.15.0
image_features = torch.cat(image_features, dim=0).to(inputs_embeds.device, inputs_embeds.dtype)
```

The dtype cast never lived at the merge site for image/video - it has always been on the assignment, and the guard's string only matched as an accidental substring of `image_features = image_features.to(...)`. Inserting `torch.cat(...)` broke that accident. `video_features` still matched, which is why only the image half failed.

Image and video features are dtype-aligned before `masked_scatter` in every release 5.5.0 through 5.15.0, so nothing is broken for users and no production code changes here. Checked separately: `_patch_gemma4_audio_feature_dtype_on_class` against real 5.15.0 finds 0 buggy / 1 fixed cast, takes the already-fixed-upstream branch and emits no drift warning; `fix_gemma4_audio_feature_dtype` correctly leaves the source unchanged.

The same release also fixed the audio bug the patch exists for (transformers #45192):

```python
# 5.5.0 - 5.14.1
audio_mask.to(inputs_embeds.device), audio_features.to(inputs_embeds.device)
# 5.15.0
audio_mask.to(inputs_embeds.device), audio_features.to(inputs_embeds.device, inputs_embeds.dtype)
```

## The change

Rather than re-anchor on the new spelling, the canaries now trace dtype ALIGNMENT structurally: parse `Gemma4Model.forward`, find the `masked_scatter` consuming the feature, and require an `inputs_embeds.dtype` cast either inside the merge argument or on the last assignment to that name before the merge (ordered by `lineno`, so a later rebind cannot be satisfied by an earlier cast).

That is strictly stronger than the substring, not a widening:

| case | should be | old | new |
|---|---|---|---|
| 5.15.0 pristine (image) | pass | FAIL | pass |
| 5.14.1 pristine (image) | pass | pass | pass |
| cast, then rebind with `.float()` | fail | fail | fail |
| dtype cast moved into a dead `if False:` branch (video) | fail | PASS, missed | fail |

Across 5.5.0, 5.10.0, 5.13.1, 5.14.0, 5.14.1 and 5.15.0 the tracer reports image/video aligned everywhere, and audio not-aligned on 5.5.0-5.14.1 (correctly, the bug unsloth patches) and aligned on 5.15.0. The audio canary now accepts "upstream aligned OR the unsloth rewriter still matches", and an ungated `test_audio_compiler_transform_still_fixes_the_device_only_spelling` keeps the rewriter under test now that 5.15 makes the real-source canary trivially green.

## Verification

All CPU-only, no GPU.

| environment | result |
|---|---|
| transformers 5.15.0, trl 1.9.2, torch 2.10.0+cpu | 27 passed (was 1 failed / 24 passed) |
| transformers 5.14.1 (old spelling) | 27 passed |
| transformers 4.57.6, trl 0.22.2, torch 2.6.0+cpu | 22 passed, 5 skipped |

Re-confirmed independently: the pre-change file still fails on 5.15.0 in the same venv, so the fix is what moves it. Pure stdlib `ast` plus pytest, so platform- and version-agnostic.

The other three failures in that CI job are untouched and tracked separately.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_gemma4_dtype_drift_guards.py [hunk 1]
```diff
@@ -195,27 +199,143 @@ def _gemma4_modeling_source():
...
+    for sub in ast.walk(node):
+        if not (
+            isinstance(sub, ast.Call)
+            and isinstance(sub.func, ast.Attribute)
+            and sub.func.attr == "to"
+        ):
+            continue
+        values = list(sub.args) + [kw.value for kw in sub.keywords]
+        if any(_is_attr(v, "inputs_embeds", "dtype") for v in values):
+            return True
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py:261]: Not taking this one. The suggestion is to also accept a bare .to(inputs_embeds.dtype) without the device, but that is exactly the drift the guard exists to catch: on a multi-device load the merge needs both, and 5.15.0 upstream spells it with both. Widening the pattern would make the guard silent on the half-fix.

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_gemma4_dtype_drift_guards.py [hunk 2]
```diff
@@ -195,27 +199,236 @@ def _gemma4_modeling_source():
...
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py:261]: Not taking this one, and one premise in it is wrong: the image/video guard does not require the device. It calls the same _merge_dtype_alignment, and there is no device requirement anywhere in the tracer - so this is not an inconsistency between two guards, it is a request to widen the one guard there is.

The tracer traces dtype alignment. Requiring inputs_embeds.device in the same .to(...) produces false reds on spellings that are device-safe. Two mutants of the real 5.15.0 source, both dtype- and device-correct, that the widened rule would fail:

  device carried at the merge argument, dtype on the assign: torch.cat(...).to(inputs_embeds.dtype) then masked_scatter(mask, image_features.to(inputs_embeds.device))  ->  (True, True) today, (True, False) widened
  dtype-only cast-if-needed, blip_2 spelling                                                                                      ->  (True, True) today, (True, False) widened

That second spelling is what upstream actually ships for this idiom: blip_2/modeling_blip_2.py:1220 if query_output.dtype != image_embeds.dtype: query_output = query_output.to(image_embeds.dtype), and granite_speech:461.

Device is not unchecked either. _gemma4_audio_merge_casts only matches a merge source whose first argument is inputs_embeds.device, and test_real_gemma4_audio_eager_patch_matches_real_merge_site runs that matcher against the real source.

The widened rule also moves no verdict on any of the six versions tested, so it buys nothing today and adds false-red risk later. I added a control test pinning the dtype-only cast-if-needed spelling with the device carried at the merge, so this stays a deliberate decision rather than an accident.

If you want the device pinned for image and video too, that is a fair ask, but it is a differently shaped check - device carried somewhere on the path to the merge, not device inside the same .to(...) as the dtype - and it belongs in its own guard.

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_gemma4_dtype_drift_guards.py [hunk 3]
```diff
@@ -195,27 +201,664 @@ def _gemma4_modeling_source():
...
+                if carries:
+                    tainted.update(bound)
+                elif node is statement:
+                    # A direct, unconditional rebinding to something the merge
+                    # did NOT flow into overwrites the name: the merged value is
+                    # gone. A rebinding nested inside a conditional may not run,
+                    # so it must not clear the taint.
+                    tainted.difference_update(bound)
+        if "inputs_embeds" in tainted:
+            return True
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py:841]: Rejecting, and the reason is that I implemented the suggested remedy and measured it making things worse.

The case reproduces exactly as written: `original = inputs_embeds; inputs_embeds = inputs_embeds.masked_scatter(...); inputs_embeds = original` traces to (True, True), a false green.

Two measurements against it. First, no gemma4 version reassigns `inputs_embeds` after the merges at all. I fetched all 23 gemma4-bearing transformers releases (5.5.0 through 5.15.0; gemma4 does not exist before 5.5.0) and checked programmatically for any post-merge assignment whose right-hand side does not reference `inputs_embeds`: none, in any version.

Second, the naive fix is not safe. Keeping the scan going and requiring `inputs_embeds` to still be tainted at the end does flip your case to (True, False) as intended, but it also flips `inputs_embeds = ...masked_scatter(...); hidden = self.language_model(inputs_embeds); inputs_embeds = None` to (True, False) - a false red on a merge that plainly did reach the model. Getting that right needs use-before-overwrite analysis, which is a lot of machinery for a shape upstream has never written.

So this is recorded as a documented limitation of the tracer rather than fixed. If upstream ever does reassign after the merge, the reachability argument changes and so does this answer.

File: unsloth_zoo/temporary_patches/gemma4.py [hunk 4]
```diff
@@ -708,6 +702,11 @@ def _patch_gemma4_audio_feature_dtype_on_class(model_cls):
...
         ):
             continue
+        if not any(
+            isinstance(inner, ast.Name) and inner.id == "inputs_embeds"
+            for inner in ast.walk(node.func.value)
+        ):
+            continue
         source = node.args[1]
         if not (
             isinstance(source, ast.Call)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4.py:723]: Rejecting on reachability plus a deliberate design choice about which way this should fail.

The asymmetry is real and reproduces: `inputs_embeds.masked_scatter(audio_mask, audio_features.to(inputs_embeds))` gives the eager matcher (0 buggy, 0 fixed) while the structural tracer reads it as aligned, so on such a refactor the canary would go red.

But I counted the spelling across every model family in transformers 5.15.0 - 503 of them - and `.to(inputs_embeds)` as a tensor-to-tensor overload appears zero times, in gemma4 or anywhere else.

More to the point, the failure mode here is loud and correct. If upstream switched to that spelling, `test_real_gemma4_audio_eager_patch_matches_real_merge_site` goes red and the runtime patch warns that it needs updating. That is exactly what the canary is for. Teaching the production matcher to silently accept a spelling we have never observed would convert a visible prompt-to-update into an invisible one, which is the worse trade. Documented as an intentional asymmetry in the test file rather than closed.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 5]
```diff
@@ -195,27 +201,664 @@ def _gemma4_modeling_source():
...
+_WITH_NODES = (ast.With, ast.AsyncWith)
+# Every construct that can rebind a plain name in the code's OWN scope. An
+# assignment statement is only the most common one: `for image_features in
+# image_batches:`, `with self.autocast() as image_features:` and
+# `if (image_features := self.pack(image_features)) is None:` all replace the
+# tensor the name refers to just as completely. Counting only assignments lets an
+# earlier aligned assignment keep proving alignment for a value that one of these
+# has since overwritten, and the replacement reaches masked_scatter un-cast.
+# (Comprehensions bind in their own scope and so cannot affect `forward`'s name.)
+_BINDING_NODES = _ASSIGN_NODES + _LOOP_NODES + _WITH_NODES + (ast.NamedExpr,)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py:496]: Rejecting on reachability. The gap is real - `ast.Match` is not in `_BINDING_NODES` and `_bound_names` has no `MatchAs`/`MatchStar` handling, so `match replacement: case [image_features]: pass` after an aligned assignment traces to (True, True).

Measured: not one `modeling_*.py` in transformers 5.15.0 contains a `match` statement at all, across all 503 model families, and none of the 23 gemma4 sources does. Supporting it means adding a new statement class to the binding model plus pattern-capture extraction for a construct upstream has never used anywhere in the library. Recorded as a documented limitation instead, so the next person reading the tracer knows it is a choice rather than an oversight.

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_gemma4_dtype_drift_guards.py [hunk 6]
```diff
@@ -195,27 +201,850 @@ def _gemma4_modeling_source():
...
+    `image_features.device` or `merged.dtype` too. A metadata read carries none
+    of the tensor's values, so the whole subtree under it is skipped; everything
+    else still counts, which keeps
+    `image_features.to(...).reshape(-1, image_features.shape[-1])` a use.
+    """
+    if isinstance(node, ast.Attribute) and node.attr in _METADATA_ATTRS:
+        return False
+    if matches(node):
+        return True
+    return any(_carries_value(child, matches) for child in ast.iter_child_nodes(node))
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py:701]: Correct about `.to(other)` semantics, but not reachable. I scanned all 434 `modeling_*.py` in transformers 5.5.0 plus 12 archived gemma4 versions: `fallback.to(image_features)` and `original.to(merged)` occur zero times. The tracer only has to classify the spellings upstream actually ships, and it classifies all 6 archived real gemma4 sources identically to main.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 7]
```diff
@@ -195,27 +201,850 @@ def _gemma4_modeling_source():
...
+    is safe", which only holds while the DESTINATION is at that dtype too. The
+    receiver may be `inputs_embeds` itself or a chain of dtype-preserving
+    transformations of it (upstream blip_2 really does write
+    `inputs_embeds.to(language_model_inputs.device).masked_scatter(...)`), but
+    `inputs_embeds.float().masked_scatter(mask, feat.to(inputs_embeds.dtype))`
+    merges a low-precision source into an fp32 destination and raises.
+    """
+    while True:
+        if isinstance(node, ast.Name):
+            return node.id == "inputs_embeds"
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py:357]: The dtype rule is right, the pattern does not exist. No `modeling_*.py` in 5.5.0, and none of the 12 archived gemma4 versions, rebinds `inputs_embeds` between the feature cast and the merge. Tracking a rebinding that never occurs would add a branch no real source exercises.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 8]
```diff
@@ -195,27 +201,850 @@ def _gemma4_modeling_source():
...
+    if isinstance(node, (ast.AnnAssign, ast.AugAssign, ast.NamedExpr)):
+        return [(node.target, getattr(node, "value", None))]
+    if isinstance(node, _LOOP_NODES):
+        return [(node.target, node.iter)]
+    if isinstance(node, _WITH_NODES):
+        return [
+            (item.optional_vars, item.context_expr)
+            for item in node.items
+            if item.optional_vars is not None
+        ]
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py:517]: Right that `with X as y` binds `__enter__()`, not `X`. There is no `with` statement anywhere in the merge path of any gemma4 version I checked, archived or current, so the tracer never reaches the case.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 9]
```diff
@@ -195,27 +201,850 @@ def _gemma4_modeling_source():
...
+    `inputs_embeds.clone().masked_scatter_(...)`, or the device-changing
+    `inputs_embeds.to(...).masked_scatter_(...)`, both of which
+    `_receiver_preserves_embeds_dtype` accepts as dtype-safe - the mutation lands
+    in a temporary and `forward` still returns the unchanged `inputs_embeds`. So
+    only skip the result-flow analysis for a receiver that is guaranteed to alias
+    the original embeddings; anything else has to hand its result back like the
+    out-of-place spelling does (`masked_scatter_` returns the receiver, so it can).
+    """
+    if call.func.attr.endswith("_") and _is_embeds_tensor(call.func.value):
+        return True
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py:802]: `detach()` does share storage, and `inputs_embeds.detach().masked_scatter_(...)` would indeed update the embeddings. It appears in no upstream source. This one also fails loudly rather than silently, so a future version that did use it would surface as a test failure rather than a missed merge.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Cross-reference before this merges, because #1051 rewrites the same helper in this same file and the two will conflict textually in `tests/test_gemma4_dtype_drift_guards.py`.

This PR is the more complete of the two: it confines the trace to `Gemma4Model.forward`, checks the receiver, reads only `args[1]`, excludes subscript targets, recognises `.to(other)`, and adds the dominator analysis, the dtype-alias fixpoint and the merge-result-is-used taint check. That covers seven of the eight items raised on #1051.

One case it does not cover, which is worth carrying over rather than losing:

`_binding_is_aligned` treats

```python
image_features = image_features.to(inputs_embeds.device)
```

as dropping dtype alignment, and it does not. `Tensor.to(device=None, dtype=None)` infers the dtype from `self`, so a device-only rebinding preserves whatever dtype the tensor already had. Confirmed against the torch docs for `torch.Tensor.to` ("If dtype is None it is inferred to be self.dtype") and empirically on torch 2.9.1.

The consequence is a false red: correct upstream source reported as drifted, which is the exact failure mode this analyzer exists to remove. It is not a correctness hole in the guard, so it does not block this PR, but whoever lands these two should make sure the device-only rebinding case survives into whichever one merges.
